app/views/users/_user.html.erb

```diff
@@ -0,0 +1,151 @@
+<div class="media">
+  <%= image_tag user.avatar_image, class: "mr-3 rounded-circle", width: 128 %>
+
+  <div class="media-body">
+    <div class="d-flex justify-content-between align-items-center">
+      <h1 class="mt-0 mb-3 display-4">
+        <%= user.username %>
+      </h1>
+
+      <div>
+        <span class="badge badge-secondary">
+          <%= user.private? ? "Private" : "Public" %>
+        </span>
+      </div>
+    </div>
+
+    <div class="row mb-4">
+      <div class="col">
+        <%=
+          link_to "#",
+            data: { 
+              toggle: "modal", 
+              target: "##{dom_id(user)}_followers"
```

Comment:
What's going on here? To start, what is `dom_id`? Domain id? Also, what is `toggle: "modal"`?

Reply:
> To start, what is dom_id?

`dom_id` is a small but handy view helper method that boils down to something like this:

```
def dom_id(active_record_object)
  "#{active_record_object.class}_#{active_record_object.id}"
end
```

So in this case, when I said `dom_id(user)`, it would return something like `"user_42"`. And `"##{dom_id(user)}_followers"` would return something like `"user_42_followers"`.

I could have pretty easily written that string concatenation myself, but Rails includes the `dom_id` helper because we end up doing this _a lot_ — we'll need to put an `id=""` attribute on a particular HTML element that corresponds to a record. Rails helper methods like `form_with` internally use `dom_id` as well — [the official doc on `dom_id` is short and shows examples](https://api.rubyonrails.org/v5.2/classes/ActionView/RecordIdentifier.html).

---

In any case, I think the real meat of the functionality/mystery here might be:

> Also, what is toggle: "modal"?

This is to do with [the Bootstrap Modal](https://getbootstrap.com/docs/4.6/components/modal), which is how I chose to display the follower/following/pending lists; I put them right on the profile page but hid them initially, and allow them to be called up with a modal, rather than show them on separate pages.

In particular, you can have multiple modals on the page, with different links bound to each. See [this example](https://getbootstrap.com/docs/4.6/components/modal/#vertically-centered); each button brings up a different modal. They both, and all the others, are in the DOM the whole time; they're just toggled visible/invisible by their buttons.

The `link_to` is just generating an `<a>` at the end of the day; view source to see the final HTML of the `<a>`. From there, it's a Bootstrap exercise.

Reply:
this is great! I was confused about the `dome_id` helper. Didn't know that it is a built-in rails helper. This can be used in a number of places! I can think of lots of modal ids that needed this unique naming convention in our app as well.